### DISCUSSION TITLE
「重学TS 2.0 」TS 面试题：declare 关键字有什么作用？

### DISCUSSION THREAD

**Discussion by semlinker** [semlinker/awesome-typescript#76] [General]
### declare 关键字有什么作用？

**Comment by wenkm** [comment]
声明全局变量的

**Comment by douhao1988** [comment]
可以通过 `declare` 关键字来告诉 `TypeScript`，你正在试图表述一个其他地方已经存在的代码，如：写在 `JavaScript` 或者是像浏览器和 `Node.js` 运行环境里的代码：
```javascript
foo = 123; // Error: 'foo' is not defined
```
和：
```typescript
declare var foo: any;
foo = 123; // allow
```
如果一个文件有扩展名 `.d.ts`，这意味着每个根级别的声明都必须以 `declare` 关键字作为前缀。这有利于让开发者清楚的知道，在这里 `TypeScript` 将不会把它编译成任何代码，同时开发者需要确保这些在编译时存在。

**Comment by s1pz** [comment]
其实我有个问题，声明全局变量不用declare也可以啊，只要没import或者export就是全局变量。所以我觉得declare的作用不是用来声明全局变量的，而是用来声明变量的，因为declare的变量是不能赋值的(Initializers are not allowed in ambient contexts)。另外.d.ts中每个根级别的声明都必须以 declare 关键字作为前缀。第三点，declare用来修饰namespace的话，namespace中的子属性都不用export 就可以直接访问。第四点，declare 在局部作用域（也必须在局部作用域）可以接global声明全局变量
